### PR TITLE
Fix Websocket Consume Messages in Partitioned Topics

### PR DESCRIPTION
**Pull request by k2la**
### Motivation

When clients consumed messages in partitioned topics on websocket, it occurred above errors:
```
15:46:31.854 [pulsar-web-28-8] WARN  o.a.pulsar.websocket.ConsumerHandler - Unhandled Error (closing connection)
java.lang.NullPointerException: null
	at org.apache.pulsar.client.impl.MultiTopicsConsumerImpl.doAcknowledge(MultiTopicsConsumerImpl.java:384) ~[pulsar-client-original-2.1.1.2-incubating-yjrelease.jar:2.1.1-incubating]
	at org.apache.pulsar.client.impl.ConsumerBase.acknowledgeAsync(ConsumerBase.java:243) ~[pulsar-client-original-2.1.1.2-incubating-yjrelease.jar:2.1.1-incubating]
	at org.apache.pulsar.websocket.ConsumerHandler.onWebSocketText(ConsumerHandler.java:238) ~[pulsar-websocket-2.1.1.2-incubating-yjrelease.jar:2.1.1-incubating]
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextMessage(JettyListenerEventDriver.java:189) ~[websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.message.SimpleTextMessage.messageComplete(SimpleTextMessage.java:69) ~[websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.appendMessage(AbstractEventDriver.java:66) ~[websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextFrame(JettyListenerEventDriver.java:158) ~[websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:162) ~[websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:376) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.extensions.AbstractExtension.nextIncomingFrame(AbstractExtension.java:176) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.extensions.compress.PerMessageDeflateExtension.nextIncomingFrame(PerMessageDeflateExtension.java:105) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.extensions.compress.PerMessageDeflateExtension.incomingFrame(PerMessageDeflateExtension.java:70) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:220) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:220) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.Parser.parse(Parser.java:256) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.readParse(AbstractWebSocketConnection.java:679) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:511) [websocket-common-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) [jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) [jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:202) [jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) [jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) [jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93) [jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) [jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_181]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_181]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-all-4.1
```

### Modifications

- Add `public` to `TopicMessageIdImpl` class.
- Parse messages and get a partitioned topic name.
- If a message is from a partitioned topic, `msgId` changes `TopicMessageIdImpl` instance.  

### Result

When clients are able to consume messages in partitioned topics on websocket.

### CODE REVIEW

**Comment by k2la** [comment]
Thank you, but I think it is difficult to get topic name in `MessageId#fromByteArray` .

**Comment by maskit** [comment]
Then how about passing it?

I read a few other code and realized that there is no other places that create `MessageId` with a topic name like this. So we don't need to worry about other places. However, this doesn't mean we are allowed to do anything here.

To be honest, I don't have great ideas. I'd make pretty big changes If I had to cleanup.  A compromise would be introducing `fromByteArrayWithTopic` or adding `Object context` parameter to `fromByteArray`. Then you can pass `topic` and do the similar thing in it. Adding context parameter is more generic but it would require `instanceof`. I'm not sure this cost is acceptable here.

Problems on current change I think are:

1) Dependency for MessageId#toString()
Since the format is not documented, there is no guarantee of the format. Only place where we can do this *parse* safely is inside `MessageIdImpl` class, MessageIdImpl have the raw values though. Other places should not depend on the format. Because of this reason I'd say current change is fragile. Also it requires many string operations.

2) Use of Impl class
Although you can find many places that use SomethingImpl class, depending on concrete classes is not a good idea in general. Since `ConsumerHandler` uses a build to create a consumer, it should not know what type of MessageId is required by the consumer exactly. Use of concrete classes makes changes difficult in long term.

**Comment by k2la** [comment]
Thank you for explaining so politely, I understand these problems.
I added `MessageId#fromByteArrayWithTopic` to solve them temporarily.

**Comment by merlimat** [comment]
run cpp tests

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TopicMessageIdImpl.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
     private final String topicName;
     private final MessageId messageId;
 
-    TopicMessageIdImpl(String topicPartitionName, String topicName, MessageId messageId) {
+    public TopicMessageIdImpl(String topicPartitionName, String topicName, MessageId messageId) {
```
Comment (maskit) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TopicMessageIdImpl.java]: Is this public still needed?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java [hunk 2]
```diff
@@ -113,6 +114,36 @@ public static MessageId fromByteArray(byte[] data) throws IOException {
...
+        }
+
+        MessageId messageId;
+        if (idData.hasBatchIndex()) {
+            messageId = new BatchMessageIdImpl(idData.getLedgerId(), idData.getEntryId(), idData.getPartition(),
+                    idData.getBatchIndex());
+        } else {
+            messageId = new MessageIdImpl(idData.getLedgerId(), idData.getEntryId(), idData.getPartition());
+        }
+        if (idData.getPartition() > -1 && topicName != null) {
```
Comment (maskit) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java]: It seems like we can merge these two if statements.

Comment (maskit) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java]: Oh, we need to pass it to the constructor... sounds odd though.

**Pull request merged**